### PR TITLE
Black Format CI task

### PR DESCRIPTION
**Pull request by AhmetCanSolak**
# Description

Black format checking task added to CI configuration, little update made on CONTRIBUTING guideline, excludes moved to `pyproject.toml`.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -1,3 +1,35 @@
+task:
```
Comment (kne42) [inline .cirrus.yml:1]: you should name these tests:
`lint_task` and `test_task`

File: .cirrus.yml [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
+  auto_cancellation: true
+
+  env:
+    matrix:
+      PY_VER: "3.6"
+      PY_VER: "3.7"
+
+  matrix:
```
Comment (kne42) [inline .cirrus.yml]: this doesn't need a matrix

File: .cirrus.yml [hunk 3]
```diff
@@ -1,3 +1,35 @@
...
+    - name: Black Formatting
+      container:
+        image: gcc:latest
```
Comment (kne42) [inline .cirrus.yml]: I would use a python image: https://cirrus-ci.org/examples/#python

File: .cirrus.yml [hunk 4]
```diff
@@ -1,3 +1,35 @@
...
+      env:
+        PATH: $HOME/conda/bin:$PATH
+      system_script:
```
Comment (kne42) [inline .cirrus.yml]: don't need this

File: .cirrus.yml [hunk 5]
```diff
@@ -1,3 +1,35 @@
+task:
+  auto_cancellation: true
+
+  env:
+    matrix:
+      PY_VER: "3.6"
```
Comment (kne42) [inline .cirrus.yml]: do we really need to check against multiple python versions?

File: .cirrus.yml (hunk 1 above)
Comment (AhmetCanSolak) [reply .cirrus.yml:1]: but that will be reduced same we have right according to docs. Still want?

File: .cirrus.yml (hunk 5 above)
Comment (AhmetCanSolak) [reply .cirrus.yml]: maybe not

File: .cirrus.yml (hunk 2 above)
Comment (AhmetCanSolak) [reply .cirrus.yml]: true, fixing

File: .cirrus.yml (hunk 3 above)
Comment (AhmetCanSolak) [reply .cirrus.yml]: didn't know about, thanks!

File: .cirrus.yml (hunk 4 above)
Comment (AhmetCanSolak) [reply .cirrus.yml]: okay removing.

File: .cirrus.yml [hunk 6]
```diff
@@ -1,3 +1,29 @@
+task:
+  auto_cancellation: true
+
+  env:
+      PY_VER: "3.6"
```
Comment (kne42) [inline .cirrus.yml]: this has no effect on the task and should be removed

File: .cirrus.yml [hunk 7]
```diff
@@ -1,3 +1,29 @@
...
+
+  name: Black Formatting
+  container:
+      image: python:latest
```
Comment (kne42) [inline .cirrus.yml]: this is where you'd specify which language version we're testing against so maybe you want to pin it to 3.6 as you did above?

File: .cirrus.yml [hunk 8]
```diff
@@ -1,3 +1,29 @@
...
+  env:
+      PATH: $HOME/conda/bin:$PATH
```
Comment (kne42) [inline .cirrus.yml]: don't need this anymore

File: .cirrus.yml [hunk 9]
```diff
@@ -1,3 +1,29 @@
...
+  conda_script:
```
Comment (kne42) [inline .cirrus.yml]: this was only here to install python but we're just using a container version so it's no longer necessary and should be removed

File: .cirrus.yml [hunk 10]
```diff
@@ -1,3 +1,29 @@
...
+      - curl https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh > install.sh
+      - bash install.sh -b -p $HOME/conda
+      - conda update -yn base conda
+      - conda install -y python=$PY_VER
+      - rm install.sh
+
+  install_script:
+    # pip >= 19.0 crashes on Linux with Python 3.7
+    - python -m pip install --retries 3 --upgrade "pip<19.0"
+    - pip install --retries 3 -r requirements.txt
```
Comment (kne42) [inline .cirrus.yml]: we really only need `` pip install `cat requirements/development.txt | grep black` ``

File: .cirrus.yml [hunk 11]
```diff
@@ -1,3 +1,29 @@
...
+
+  test_script:
+    - black -S -l 79 -t py36 --exclude vendored\|_vispy --check .
```
Comment (kne42) [inline .cirrus.yml]: shouldn't all of these command line options (except `--check`) already be set by the `pyproject.toml` (and therefore autorun when we `black --check`)?

File: .cirrus.yml (hunk 1 above)
Comment (kne42) [reply .cirrus.yml:1]: ah, you're right, nvm

File: .cirrus.yml (hunk 11 above)
Comment (AhmetCanSolak) [reply .cirrus.yml]: nope, those are for pre-commit

Comment (kne42) [reply .cirrus.yml]: https://github.com/python/black#pyprojecttoml

Comment (AhmetCanSolak) [reply .cirrus.yml]: yeah but we have excludes and stuff, also that line matches our contributing guideline?

Comment (kne42) [reply .cirrus.yml]: our contributing guidelines are there to tell people what it's equivalent to but i would prefer to cut down on redundancy wherever possible

**Comment by AhmetCanSolak** [comment]
@kne42 i addressed your concern, can you provide a new review?

File: .cirrus.yml [hunk 12]
```diff
@@ -1,3 +1,18 @@
+task:
+  auto_cancellation: true
+
+  name: Black Formatting
+  container:
+      image: python3.6:latest
```
Comment (kne42) [inline .cirrus.yml]:
```suggestion
      image: python:3.6
```

File: .cirrus.yml [hunk 13]
```diff
@@ -1,3 +1,18 @@
...
+
+  install_script:
+    # pip >= 19.0 crashes on Linux with Python 3.7
+    - python -m pip install --retries 3 --upgrade "pip<19.0"
+    - pip install `cat requirements/development.txt | grep black`
+
+  test_script:
+    - black --exclude vendored\|_vispy --check .
```
Comment (kne42) [inline .cirrus.yml]: could you add this `--exclude` stuff to `pyproject.toml` and remove it from here? you can find an example setup [here](https://github.com/python/black/blob/master/pyproject.toml#L12)

File: pyproject.toml [hunk 14]
```diff
@@ -2,3 +2,4 @@
 target-version = ['py36', 'py37']
 skip-string-normalization = true
 line-length = 79
+exclude = 'vendored\|_vispy'
```
Comment (kne42) [inline pyproject.toml]: test is failing because this somehow did not properly exclude the desired directories

Comment (AhmetCanSolak) [reply pyproject.toml]: yeah same command line string not working

**Review by kne42 — APPROVED** [review]
Sweet, looks great! 🎉

**Pull request merged**